### PR TITLE
Added Firestore type parameters

### PR DESCRIPTION
**Pull request by Phaen**
### Summary

Whenever `DocumentSnapshot.get` was used to retrieve a document field, the user would have to deal with the type assertion themselves, leading to ugly `as` statements throughout their code. Additionally, the typing might not be clear to the user, as firestore accepts `Date` as a field type when storing data, but would return it as a `FirebaseFirestoreTypes.Timestamp` upon retrieval. By allowing the user to pass the type and constricting the type parameter, it should help alleviate these issues.

### Checklist

- [x] Supports `Android`
- [x] Supports `iOS`
- [ ] `e2e` tests added or updated in packages/**/e2e
- [ ] Flow types updated
- [x] Typescript types updated

### Release Plan

[TYPES] [BREAKING] [FIRESTORE] - Added type parameters to `DocumentSnapshots.get`

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Donate via [Open Collective](https://opencollective.com/react-native-firebase/donate)
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter
- 👉 Star this repo on GitHub ⭐️
- 👉 Contribute; see our [contributing guide](/CONTRIBUTING.md)

### CODE REVIEW

File: packages/firestore/lib/index.d.ts [hunk 1]
```diff
@@ -494,12 +512,12 @@ export namespace FirebaseFirestoreTypes {
      * ```js
      * const user = await firebase.firestore().doc('users/alovelace').get();
      *
-     * console.log('Address ZIP Code', user.get('address.zip'));
+     * console.log('Address ZIP Code', user.get<string>('address.zip'));
```
Comment (russellwheatley) [inline packages/firestore/lib/index.d.ts]: you wouldn't put `<string>` in the actual code.

Comment (russellwheatley) [reply packages/firestore/lib/index.d.ts]: To clarify, it presumes the developer is using Typescript.

File: packages/firestore/lib/index.d.ts [hunk 2]
```diff
@@ -175,6 +175,24 @@ export namespace FirebaseFirestoreTypes {
    */
   export type DocumentChangeType = 'added' | 'removed' | 'modified';
 
+  /**
+   * The types for a DocumentSnapshot field that are supported by Firestore.
+   */
+  export type DocumentFieldType =
+    | string
```
Comment (russellwheatley) [inline packages/firestore/lib/index.d.ts]: Apologies. A couple of other things that I missed. the `|` operator shouldn't be on line 182.

File: packages/firestore/lib/index.d.ts [hunk 3]
```diff
@@ -499,7 +517,7 @@ export namespace FirebaseFirestoreTypes {
      *
      * @param fieldPath The path (e.g. 'foo' or 'foo.bar') to a specific field.
      */
-    get(fieldPath: string | FieldPath): any;
+    get<fieldType extends DocumentFieldType>(fieldPath: string | FieldPath): fieldType;
```
Comment (russellwheatley) [inline packages/firestore/lib/index.d.ts]: Shouldn't this just be `get(fieldPath: string | FieldPath): DocumentFieldType;`? `fieldType` doesn't actually extend `DocumentFieldType` with any further properties

Comment (Phaen) [reply packages/firestore/lib/index.d.ts]: It defines a new named type to be specifically repeated for the returning value. Is there a different way that you're supposed to do that?

File: packages/firestore/lib/index.d.ts (hunk 2 above)
Comment (Phaen) [reply packages/firestore/lib/index.d.ts]: This operator is indeed redundant, but harmless. It's automatically generated by Prettier for consistent styling.

File: packages/firestore/lib/index.d.ts (hunk 1 above)
Comment (Phaen) [reply packages/firestore/lib/index.d.ts]: I forgot that we're not using Typescript in any of the examples, good one!

**Pull request merged**